src/ULog.ts

```diff
@@ -220,37 +284,119 @@ export class ULog {
     return this.#dataTimeRange;
   }
 
-  async #createIndex(): Promise<void> {
+  async #createIndex(reader: ChunkedReader): Promise<void> {
+    if (!this.#header || this.#dataEnd == undefined) {
+      throw new Error(`Cannot read before open`);
+    }
+
+    const dataEnd = this.#dataEnd;
     const timeIndex: IndexEntry[] = [];
     const dataCounts = new Map<number, number>();
     let minTimestamp: bigint | undefined;
     let maxTimestamp = 0n;
     let logMessageCount = 0;
 
-    let i = 0;
-    let message: DataSectionMessage | undefined;
-    while ((message = await this.#readParsedMessage())) {
-      if (message.type === MessageType.Data) {
-        if (minTimestamp == undefined || message.value.timestamp < minTimestamp) {
-          minTimestamp = message.value.timestamp;
+    // The offset of the timestamp field for a data message. The offset is from the start of the
+    // message data.
+    const timestampFieldOffsets = new Map<MsgId, number>();
+
+    for (;;) {
+      const offset = reader.position();
+
+      // If there is less than 3 bytes left in the file, we can't read a message header so we're done
+      if (dataEnd - offset < 3) {
```

Comment:
Is `0 < dataEnd - offset < 3` allowed for a valid Ulog file?

Reply:
I don't understand. Allowed in what sense? A header is 3 bytes and comes before each record so if there's no longer 3 bytes left to read then we are at the end.

Reply:
What if there are 2 bytes to read still? Does that mean the file is somehow corrupted or wasn't closed properly?

Reply:
Don't know - haven't investigated that part. We had this logic before when reading so I kept it.